### DISCUSSION TITLE
Is it possible to reuse a token when the user comes back before token expires ?

### DISCUSSION THREAD

**Discussion by BobForCat** [keycloak/keycloak#15460] [Q&A]
Hi

I'll explain the behaviour I understand now, and what I'd suggest 

Here's what happen now :

1. the user open my website, written in ReactJS
2. he got redirected to my keycloak, it authenticates him using kerberos available at my company network/laptops
3. he got redirected to my website, I store the token (to be given to APIs) and the refreshtoken in the localStorage

Then the use goes away and come back like 2 minutes later, or just refreshes the page => that retriggers the scenario written just above, the whole redirect stuff

As the user has logged-in in less time than the token validation limit and the token I saved is still valid regarding a client-API I'd call, why isn't it possible that a new session would try to **use that previous information** (token or else) to be already ready and **avoid that 1-2sec redirection** ? 

I hope that's clear, and I'm here to provide more details if they are in my scope (I'm only a developper, and the IT guys haven't been able to answer me)

Thanks

**Comment by DAHAG-ArisNourbakhsh** [comment]
Maybe I'm grossly misunderstanding your flow, but to me it reads like that the decision to redirect to keycloak despite having a valid accessToken laying around in the localstorage is 100% under your control.

Just don't redirect the user in that scenario.

**Reply by BobForCat** [reply]
I'm not doing the redirection myself, I instanciate the keycloak, which redirects to the keycloak url which then redirect back to the app

Here's a minimal code of mine to show what I'm doing

```js
import * as React from "react";
import Keycloak from "keycloak-js";

export class App extends React.Component<any, any> {
    constructor(props: any) {
        super(props);
        this.state = {
            token: '',
            refreshToken: '',
            keycloak: '',
            isAuthenticated: false,
        }
    }

    render(): React.ReactNode {
        const {isAuthenticated} = this.state;
        if (!isAuthenticated) { return <p>Unauthenticated</p>  }
        return (<p>OK</p>);
    }

    componentDidMount() {
        let keycloak = new Keycloak({url: "https://my-keycloak", realm: "my-realm", clientId: "my-clientId-public"});
        keycloak.init({onLoad: "login-required"});
        keycloak.onReady = (auth: boolean) => {
            keycloak.loadUserProfile().then((data: any) => {
                localStorage.setItem('my-app-token', keycloak.token || ""); // saved to be sent in axios calls to the APIs
                this.setState({
                    isAuthenticated: auth,
                    token: keycloak.token || "",
                    refreshToken: keycloak.refreshToken || "",
                    keycloak: keycloak,
                });
            });
        };
        keycloak.onTokenExpired = () => this.refreshUserToken()
    };

    refreshUserToken = () => {
        let {keycloak} = this.state;
        keycloak.updateToken(5).success((refreshed: boolean) => {
            if (refreshed) this.setState({token: keycloak.token, refreshToken: keycloak.refreshToken, keycloak: keycloak},
                () => localStorage.setItem('my-app-token', keycloak.token));
        }).error(() => console.log("Failed refresh"))
    }
}
```

**Comment by lexcao** [comment]
Hi, could you please try `check-sso` option and silent login callback with login status iframe enabled?

And when the silent check fails, you could redirect to keycloak manually by `login()` method.
Otherwise, the user could be silent login via the iframe silently.

You can refer to the doc [here](https://www.keycloak.org/docs/latest/securing_apps/index.html#_javascript_adapter)

**Reply by kingyue737** [reply]
But according to the doc, `silent-check-sso` does not work for modern browsers
